src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const currentState = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        addProperties(currentState, stateCopy, action.extraData);
+        break;
+      case 'removeProperties':
+        removeProperties(currentState, stateCopy, action.keysToRemove);
+        break;
+      case 'clear':
+        clear(currentState, stateCopy);
+        break;
+      default:
+        break;
+    }
```

Comment:
The `switch` statement correctly handles different action types. However, the code inside each case is calling functions that mutate the shared `stateCopy`. Instead, the operations to create new state objects should be done within the cases directly, avoiding mutations and function calls that handle state mutations.